mono/metadata/file-mmap-windows.c

```diff
@@ -413,7 +454,10 @@ mono_mmap_unmap (void *mmap_handle, MonoError *error)
 
 	MmapInstance *h = (MmapInstance *) mmap_handle;
 
-	gboolean result = UnmapViewOfFile (h->address);
+	gboolean result;
+	MONO_ENTER_GC_SAFE;
+	result = UnmapViewOfFile (h->address);
+	MONO_EXIT_GC_SAFE;
 
 	g_free (h);
```

Comment:
Doesn't `g_free` potentially lock also?

Reply:
> Doesn't `g_free` potentially lock also?

We're... not really prepared to deal with malloc/free properly.

Reply:
Notice that just getting the policy, at least the first time, can `g_free`. So there is a possible recursion there, if implemented.

Reply:
Ergh, yeah I think so, though I think it'll almost always be super brief similar to ReleaseSemaphore?

Reply:
`g_free` at least on Windows, will tend not to lock.
There are a few imaginable things it could do.

1. It could free to an interlocked list based on the size of the block.
Very fast, no lock (aside from what the CPU does, or an InterlockedCompareExchange loop).

2. It could coalesece adjacent free blocks into larger blocks.
I'm not sure that requires locking, or if it happens, but it probably happens.
Splitting/coalescing seems like basic heap operations, unless somehow optimized away.

3. It could, with a lot of free'ing, actually VirtualFree a large block, incuring a syscall and more locking. Though I think by default, on Windows, this might never happen.
Even if it happens sometimes, it'd be less often than you might expect, like to avoid "hysterisis" where when you are on the boundary, it'd keep going back and forth, VirtualAlloc/VirtualFree, turning a small alloc/free loop into slow.

Reply:
But anyway, it seems we are ignoring `malloc` / `free` for now.

Reply:
How would `g_free` lock?

Reply:
`g_free` and `g_malloc` are managing a per-process heap.
You can `malloc` on one thread, and `free` it on another.
At some conceptual level, it must lock, because it is managing a process-wide resource.

That said, it is an area with a lot of optimization.
There tend to be a few size-buckets and per-thread free lists.
Or interlocked process free lists.

As well, under heavy `malloc` (not necessarily `free`), where you are hitting new high water marks in terms of memory use, `malloc` must go to the kernel for more memory, `VirtualAlloc` or `mmap`. Which is a very heavyweight operation. It has to change page table entries. And flush per-processor page table caches (TLB entries).